### DISCUSSION TITLE
OptOut of Stored Procedures?

### DISCUSSION THREAD

**Discussion by StarGazer202424** [thepirat000/Audit.NET#699] [Q&A]
Hi

I have some Stored Procedures and I do not want to log into my audit.log. Most of them are just grabbing predefined lists. 

Some of these SP are being called using EF Core and some are using Ado.net.

Is there something built in to filter these out like I can do with "optout" on classes?

**Comment by thepirat000** [comment]
Are you referring to events generated with the Audit Command Interceptor?

**Reply by StarGazer202424** [reply]
I am not too sure I have stuff like this

```
using(var context = dbFactory.CreateDbContext())
{
      var picklist = context.PickList.FromSqlRaw("EXEC GetPickList).ToList()
}
```
I will aslo have ones like this but just written in Ado.net 

I see  something like this in my logs

```
{
   CommandEvent: {
         CommandText: "EXEC GetPickList)"
         Result: {
               "Table1": [
                        { "Value": "Item1"},
                         { "Value": "Item2"}
                 ]

          }
   }
}

```
I am wondering if I can somehow disable this from being recorded. I know with the EF core addon I could just go to my "Picklist" class and put "OptOut/In" and I think it would not record it.

**Reply by StarGazer202424** [reply]
I do have AuditCommand

```
build.AddInterceptors(new AuditCommandINterceptor() 
{
           AuditEventType = "EF_Command",
          ExldueNonQueryEvents = true,
         IncludeReaderResults = true,
          LogParamaterValues = true
}).EnableSenstiveDataLogging();
```

**Comment by thepirat000** [comment]
The command interceptor has no knowledge of the .NET entity classes, so there's no way to opt out by class.

However, when using the AuditCommandInterceptor, you have additional options to filter or discard events.

1. Filter the reader events with the [`IncludeReaderEventsPredicate` setting](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#settings-low-level-interceptor). For example:

```c#
optionsBuilder.AddInterceptors(new AuditCommandInterceptor()
{
    IncludeReaderEventsPredicate = cmd => cmd.CommandSource != CommandSource.FromSqlQuery || !cmd.Command.CommandText.Contains("GetPickList")
});
```

2. Discard the audit scope in the `OnScopeCreated` event with a [custom action](https://github.com/thepirat000/Audit.NET/tree/master?tab=readme-ov-file#custom-actions). For example: 

```c#
Audit.Core.Configuration.AddCustomAction(ActionType.OnScopeCreated, scope =>
{
    var cmdEvent = scope.GetCommandEntityFrameworkEvent();
    if (cmdEvent?.CommandSource == CommandSource.FromSqlQuery && cmdEvent.CommandText.Contains("GetPickList"))
    {
        scope.Discard();
    }
});
```

**Reply by StarGazer202424** [reply]
Hey, okay that is along the lines I thought I would have to do. I will try to do that and see how it goes and report back if I got any issues.

Thanks

**Reply by StarGazer202424** [reply]
@thepirat000 - Hey, your helping me with my other issue and noticed in your example project has

    `     IncludeReaderEventsPredicate = cmdData => cmdData.CommandSource == CommandSource.LinqQuery,`
         
  How do I incorporate this with what you posted here?

**Reply by thepirat000** [reply]
I don't understand your last question. The IncludeReaderEventsPredicate is a custom predicate on the commend to determine if it should be logged. I just put an example but you can add your custom logic as needed. Just set the IncludeReaderEventsPredicate property when instantiating the AuditCommandInterceptor to return true for the commands you want to log and false to ignore them

**Reply by StarGazer202424** [reply]
Sorry, in our discussion here you posted a sample project which I copied the code from into my application.

https://github.com/thepirat000/Audit.NET/discussions/692

I noticed in the sample project you have already set the "IncludeReaderEventsPredicate" I was not 100% sure what it is doing and how it ties into this problem I am having and if I need to do both. 


        ```
    // NOTE: Configure your database connection and audit interceptor
            builder.Services.AddDbContextFactory<PeopleContext>(opt =>
                opt.UseSqlite($"Data Source={nameof(PeopleContext)}.db")
                    .AddInterceptors(new AuditCommandInterceptor()
                    {
                        AuditEventType = "EF_COMMAND",
                        ExcludeNonQueryEvents = true,
                        ExcludeScalarEvents = true,
                        IncludeReaderEventsPredicate = cmdData => cmdData.CommandSource == CommandSource.LinqQuery,
                        IncludeReaderResults = true,
                        LogParameterValues = true
                    })
                    .EnableSensitiveDataLogging(), ServiceLifetime.Scoped);

            // NOTE: Configure your audit output, by default logs to files in the C:\Logs directory
            Audit.Core.Configuration.Setup()
                .IncludeActivityTrace()
                .UseFileLogProvider(f => f
                    .Directory(@"C:\Logs")
                    .FilenameBuilder(ev => $"{ev.StartDate:yyyyMMddHHmmssffff}_{ev.EventType}.json"));

```

**Reply by thepirat000** [reply]
Refer to the documentation on the [CommandSource Enum](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.diagnostics.commandsource?view=efcore-8.0#fields).

- `cmd.CommandSource == CommandSource.LinqQuery` implies: Log any query executed on a DbSet, such as `dbContext.People.First(p => p.Id == 1)`.
- `cmd.CommandSource != CommandSource.FromSqlQuery || !cmd.Command.CommandText.Contains("GetPickList")`, which is equivalent to 
`!(cmd.CommandSource == CommandSource.FromSqlQuery && cmd.Command.CommandText.Contains("GetPickList"))` means: Exclude from logging any commands executed via `FromSqlRaw` if they contain the string "GetPickList."

To combine both conditions, the predicate would be:

```c#
cmd.CommandSource == CommandSource.LinqQuery || !(cmd.CommandSource == CommandSource.FromSqlQuery && cmd.Command.CommandText.Contains("GetPickList"))
```

**Reply by StarGazer202424** [reply]
Okay let me review the link you sent. As I got a bit confused as I noticed that "GetPickList" was no longer being logged anymore when I added the line of 

`  IncludeReaderEventsPredicate = cmdData => cmdData.CommandSource == CommandSource.LinqQuery,`

So it made me wonder if this was another way to filter them out or if something else is going on.
